registry/extension/oras/artifactgarbagecollectionhandler.go

```diff
@@ -0,0 +1,191 @@
+		gc.artifactMarkSet[dgst] += 1
+
+		// mark the artifact blobs
+		descriptors := manifest.References()
+		for _, descriptor := range descriptors {
+			gc.artifactMarkSet[descriptor.Digest] += 1
+			fmt.Printf("%s: incrementing artifact blob ref count %s\n", repository.Named().Name(), descriptor.Digest)
+		}
+		return false, nil
+	} else {
+		// if the manifest passed isn't an an artifact -> call the sweep ingestor
+		// find all artifacts linked to manifest and add to artifactManifestIndex for subsequent deletion
+		gc.artifactManifestIndex[dgst] = make([]digest.Digest, 0)
+		err := enumerateReferrerLinks(ctx,
+			rootPath,
+			storageDriver,
+			repository,
+			blobStatter,
+			dgst,
+			gc.artifactManifestIndex,
+			artifactSweepIngestor)
+
+		if err != nil {
+			switch err.(type) {
+			case driver.PathNotFoundError:
+				return true, nil
+			}
+			return true, err
+		}
+		return true, nil
+	}
+}
+
+func (gc *orasGCHandler) RemoveManifest(ctx context.Context, storageDriver driver.StorageDriver, registry distribution.Namespace, dgst digest.Digest, repositoryName string) error {
+	referrerRootPath := referrersLinkPath(repositoryName)
+	fullArtifactManifestPath := path.Join(referrerRootPath, dgst.Algorithm().String(), dgst.Hex())
+	dcontext.GetLogger(ctx).Infof("deleting manifest ref folder: %s", fullArtifactManifestPath)
+	err := storageDriver.Delete(ctx, fullArtifactManifestPath)
+	if err != nil {
+		if _, ok := err.(driver.PathNotFoundError); !ok {
+			return err
+		}
+	}
+
+	subjectLinkedArtifacts, ok := gc.artifactManifestIndex[dgst]
+	if ok {
+		for _, artifactDigest := range subjectLinkedArtifacts {
+			// get the artifact manifest
+			named, err := reference.WithName(repositoryName)
+			if err != nil {
+				return fmt.Errorf("failed to parse repo name %s: %v", repositoryName, err)
+			}
+			repository, err := registry.Repository(ctx, named)
+			if err != nil {
+				return fmt.Errorf("failed to construct repository: %v", err)
+			}
+
+			manifestService, err := repository.Manifests(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to construct manifest service: %v", err)
+			}
+			artifactManifest, err := manifestService.Get(ctx, artifactDigest)
+			if err != nil {
+				return fmt.Errorf("failed to get artifact manifest: %v", err)
+			}
+
+			// extract the reference
+			blobs := artifactManifest.References()
+
+			// decrement refcount for the blobs' digests and the manifest digest
+			gc.artifactMarkSet[artifactDigest] -= 1
+			fmt.Printf("%s: decrementing artifact manifest ref count %s\n", repositoryName, dgst)
+			for _, descriptor := range blobs {
+				gc.artifactMarkSet[descriptor.Digest] -= 1
+				fmt.Printf("%s: decrementing artifact blob ref count %s\n", repositoryName, descriptor.Digest)
+			}
+			// delete each artifact manifest's revision
+			manifestPath := referrersRepositoriesManifestRevisionPath(repositoryName, artifactDigest)
```

Comment:
This should be just the manifest's revision path... can it be constructed using the already defined paths package?

Reply:
Discussed offline: we should be using the paths package however this is defined in storage package and is left unexported. Leaving as a future TODO